docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+    "event_start_time": ...,
+    "event_stop_time": ...,
+    "value": 12345454,
+}
+```
+
+I'd also suggest reporting one value, `cpu time`*`compute_units_count`, instead of two separate fields as it makes event schema simpler (it is possible to treat it the same way as traffic) and preserves additivity.
+
+## Pageserver
+
+Pageserver knows / has access to / can calculate the rest of the metrics:
+
+* Written size -- that is basically `last_received_lsn`,
+* Synthetic storage size -- there is a way to calculate it, albeit a costly one,
+* Real storage size -- there is a way to calculate it using a layer map or filesystem,
+* S3 storage size -- can calculate it by S3 API calls
+
+Some of those metrics are expensive to calculate, so the reporting period here is driven mainly by implementation details. We can set it to, for example, once per hour. Not a big deal since the pageserver is stateful, and all metrics can be reported as an absolute value, not increments. At the same time, a smaller reporting period improves UX, so it would be good to have something more real-time.
+
+`written size` is primarily a safekeeper-related metric, but since it is available on both pageserver and safekeeper, we can avoid reporting anything from the safekeeper.
+
+Example event:
+
+```json
+{
+    "metric": "s3_tenant_size_bytes",
+    "type": "absolute",
+    "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+    "time": ...,
+    "value": 12345454,
+}
+```
+
+# Data collection
+
+## Push vs. pull
+
+We already have pull-based Prometheus metrics, so it is tempting to use them here too. However, in our setup, it is hard to tell when some metric changes. For example, garbage collection will constantly free some disk space over a week, even if the project is down for that week. We could also iterate through all existing tenants/branches/endpoints, but that means some amount of code to do that properly and most likely we will end up with some per-metric hacks in the collector to cut out some of the tenants that are surely not changing that metric.
+
+With the push model, it is easier to publish data only about actively changing metrics -- pageserver knows when it performs s3 offloads, garbage collection and starts/stops consuming data from the safekeeper; proxy knows about connected clients; console / autoscaler-agent knows about active cpu time.
+
+Hence, let's go with a push-based model.
+
+## Common bus vs. proxying through the console
+
+We can implement such push systems in a few ways:
+
+a. Each component pushes its metrics to the "common bus", namely segment, Kafka, or something similar. That approach scales well, but it would be harder to test it locally, will introduce new dependencies, we will have to distribute secrets for that connection to all of the components, etc. We would also have to loop back some of the events and their aggregates to the console, as we want to show some that metrics to the user in real-time.
```

Comment:
Why didn't you describe the solution with Kafka? It's possible to achieve better reliability with transactional outbox, DLQ. And we can build windows on top of Kafka for customers to show their metrics per day/5 minutes/ or how they want via kSQL and streams/tables (taking into account that 
@arssher
 is working on incorporating protobuf, which will also be helpful here).

Reply:
I kind of assumed implicit knowledge about our segment pipeline here, but rather should have written that explicitly. So the systems of record here are segment (for recent events) and Snowflake where this event are collected in a big table.

So we can have that windows with existing systems without introducing kafka

Reply:
segment and snowflake are not presented in this schema. Nonetheless, how can we be sure that the data was delivered durably to the Snowflake?

Reply:
<img width="1266" alt="Screen Shot 2022-11-22 at 16 27 46" src="https://user-images.githubusercontent.com/284219/203339541-4bace522-7560-4afd-949d-ba3459a37594.png">

Reply:
> Nonetheless, how can we be sure that the data was delivered durably to the Snowflake?

Segment is basically managed Kafka sending data to the Snowpipe. I hope they did integration properly, but can't guarantee it as well. At the same time don't have too big concerns either -- segment will store event for a few day if can't connect to Snowflake. We can iterate on that one more time and have some system of record on our premises.